caom2pipe/client_composable.py

```diff
@@ -0,0 +1,665 @@
+#
+#  : 4 $
+#
+# ***********************************************************************
+#
+
+"""
+This module contains classes and functions that have to do with managing
+and using long-term connections to CADC services.
+
+Leave out:
+- get_artifact_metadata_client, since it mostly CAOM2 dependent, and the
+  only client-based part is get_cadc_meta_client, which is already here
+- should Metrics be here? Maybe.
+"""
+
+import logging
+import os
+import traceback
+
+from dataclasses import dataclass
+from datetime import datetime
+from io import BytesIO, StringIO
+from sys import getsizeof
+
+from astropy.table import Table
+
+from cadcdata import CadcDataClient
+from cadctap import CadcTapClient
+from cadcutils import net, exceptions
+from caom2pipe import manage_composable as mc
+from caom2repo import CAOM2RepoClient
+from vos import Client
+
+__all__ = [
+    'client_get',
+    'client_put',
+    'client_put_fqn',
+    'ClientCollection',
+    'current',
+    'data_get',
+    'data_put_fqn',
+    'declare_client',
+    'define_subject',
+    'get_cadc_headers_client',
+    'get_cadc_meta_client',
+    'get_cadc_meta_client_v',
+    'look_pull_and_put',
+    'look_pull_and_put_v',
+    'query_tap_client',
+    'repo_create',
+    'repo_delete',
+    'repo_get',
+    'repo_update',
+]
+
+
+class ClientCollection(object):
+    """
+    This class initializes and provides accessors to known HTTP clients
+    for CADC services.
+
+    This is time-dependent cohesiveness. The clients should all be declared
+    at approximately the same time, will live as long as the pipeline
+    instance, and require the same AAI credential information, even though
+    there is one per not-related-at-all CADC service.
+
+    Eventually session retries, etc, might be configurable, but that need has
+    not yet been demonstrated.
+    """
+
+    def __init__(self, config):
+        self._metadata_client = None
+        self._data_client = None
+        self._query_client = None
+        self._metrics = None
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self._init(config)
+
+    @property
+    def data_client(self):
+        return self._data_client
+
+    @property
+    def metadata_client(self):
+        return self._metadata_client
+
+    @property
+    def metrics(self):
+        return self._metrics
+
+    @property
+    def query_client(self):
+        return self._query_client
+
+    def _init(self, config):
+        subject = define_subject(config)
+
+        if mc.TaskType.SCRAPE in config.task_types:
+            self._logger.info(
+                f'SCRAPE\'ing data - no clients will be initialized.'
+            )
+        else:
+            self._metadata_client = CAOM2RepoClient(
+                subject, config.logging_level, config.resource_id
+            )
+
+            if config.features.supports_latest_client:
+                self._logger.warning('Using vos.Client for storage.')
```

Comment:
It's logged as a warning so that it's easy to tell, even when the logging level is set to `--quiet`, which client is being used. The message itself should say `StorageInventoryClient` now.